controllers/virtualmachinegrouppublishrequest/virtualmachinegrouppublishrequest_controller.go

```diff
@@ -95,63 +127,308 @@ func NewReconciler(
+	for _, req := range reqs.Items {
+		if metav1.IsControlledBy(&req, ctx.VMGroupPublishRequest) {
+			existReqsMap[req.Spec.Source.Name] = req
+			if req.Status.Ready {
+				completedReqsSet.Insert(req.Spec.Source.Name)
+			} else {
+				pendingReqsSet.Insert(req.Spec.Source.Name)
+			}
+		}
+	}
+
+	return existReqsMap, pendingReqsSet, completedReqsSet, nil
+}
+
+// reconcileStatusImages updates status's image status by taking a map of VMGroupPublishRequests
+// own by the VirtualMachineGroupPublishRequest and converting them into a sorted list of
+// VirtualMachineGroupPublishRequestImageStatus. Sorting prevents reconciling when list is unchanged.
+func (r *Reconciler) reconcileStatusImages(
+	ctx *pkgctx.VirtualMachineGroupPublishRequestContext,
+	existReqsMap map[string]vmopv1.VirtualMachinePublishRequest) {
+
+	imageStatuses := make([]vmopv1.VirtualMachineGroupPublishRequestImageStatus, 0, len(existReqsMap))
+
+	for _, req := range existReqsMap {
+		imageStatuses = append(imageStatuses, vmopv1.VirtualMachineGroupPublishRequestImageStatus{
+			Source:             req.Spec.Source.Name,
+			PublishRequestName: req.Name,
+			ImageName:          req.Status.ImageName,
+			Conditions:         req.Status.Conditions,
+		})
+	}
+
+	sort.Slice(imageStatuses, func(i, j int) bool {
+		return imageStatuses[i].Source < imageStatuses[j].Source
+	})
+
+	ctx.VMGroupPublishRequest.Status.Images = imageStatuses
+}
+
+// reconcileNewPublishRequests compares a set of exist requests and a set of expected requests and
+// creates new requests if they are missing in the exist requests set.
+// It returns nil ONLY when no new publish requests were created, in other words, the set of exist requests
+// and the set of expected requests are the same.
+func (r *Reconciler) reconcileNewPublishRequests(
+	ctx *pkgctx.VirtualMachineGroupPublishRequestContext,
+	existReqsSet, desiredReqsSet sets.Set[string]) error {
+
+	newReqsSet := desiredReqsSet.Difference(existReqsSet)
+	if newReqsSet.Len() == 0 {
+		return nil
+	}
+
+	vmGroupPubReqGvk, err := apiutil.GVKForObject(ctx.VMGroupPublishRequest, r.Scheme())
+	if err != nil {
+		return err
+	}
+
+	for sourceVM := range newReqsSet {
+		// TTLSecondsAfterFinished is intentionally unset to avoid vm publish request to be
+		// deleted after completion.
+		if err := r.Create(ctx, &vmopv1.VirtualMachinePublishRequest{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace:    ctx.VMGroupPublishRequest.Namespace,
+				GenerateName: ctx.VMGroupPublishRequest.Name + "-",
+				OwnerReferences: []metav1.OwnerReference{
+					*metav1.NewControllerRef(ctx.VMGroupPublishRequest, vmGroupPubReqGvk),
+				},
+				Labels: map[string]string{
+					vmopv1.VirtualMachinePublishRequestManagedByLabelKey: ctx.VMGroupPublishRequest.Name,
+				},
+			},
+			Spec: vmopv1.VirtualMachinePublishRequestSpec{
+				Source: vmopv1.VirtualMachinePublishRequestSource{
+					Name:       sourceVM,
+					Kind:       "VirtualMachinePublishRequest",
```

Comment:
This should be the `VirtualMachine` kind right?

Reply:
great catch. thanks!